app/models/king.rb

```diff
@@ -1,22 +1,46 @@
 class King < Piece
+  def valid_move?(new_x, new_y)
+    # return false if is_obstructed?(new_x.to_i, new_y.to_i)
+    return true if
+      x_difference == (new_x.to_i - x_position).abs
+    y_difference = (new_y.to_i - y_position).abs
+    (x_difference <= 1) && (y_difference <= 1)
+  end
+  false
+end
 
+def obstructed_squares(new_x, new_y)
+  x_diff = (new_x - x_position)
 
-	def valid_move?(new_x, new_y)
-		#return false if is_obstructed?(new_x.to_i, new_y.to_i)
-  	return true if 
-  		x_difference = (new_x.to_i - x_position).abs
-    	y_difference = (new_y.to_i - y_position).abs
-    	(x_difference <= 1) && (y_difference <= 1)
+  case x_diff
+  when 2
+    # move is kingside castle
+    rectilinear_obstruction_array(7, new_y)
+  when -2
+    # move is queenside castle
+    rectilinear_obstruction_array?(0, new_y)
+  else
+    # otherwise king moves one space - can't be obstructed - return empty
+    return []
   end
-  	false
-end 
-  #still need to account for invalid move if King moves into check
+end
 
+# determine if king can move himself out of check
+def can_move_out_of_check?
+  start_x = x_position
+  start_y = y_position
+  success = false
+  ((x_position - 1)..(x_position + 1)).each do |x|
+    ((y_position - 1)..(y_position + 1)).each do |y|
+      update_attributes(x_position: x, y_position: y) if valid_move?(x, y)
```

Comment:
@ltsinh methods with `?` are usually expected to return true or false and not have any side effects.. this one seems to change the position... maybe we can split the logic into 2 separate methods, one checking if it can move out of check and the other one doing what line 35 or 40 do

Reply:
I'll work on this refactor. And do another pull request hopefully by Sat.